pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,348 @@
+		return nil
+	}
+
+	switch m.Type {
+	case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+		return r.traceIfCovered(m)
+	case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp, raftpb.MsgStorageApplyResp:
+		r.traceIfPast(m)
+		return nil
+	}
+	return nil
+}
+
+// FlushAll will unregister all the currently active traces. It is safe to call
+// multiple times, but should always be called when the replica is destroyed.
+func (r *RaftTracer) FlushAll() {
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		r.m.Delete(index)
+		r.numRegistered.Add(-1)
+		t.logf(2, "cleanup log index %d during close", index)
+		return true
+	})
+}
+
+func peer(p raftpb.PeerID) redact.SafeString {
+	return redact.SafeString(raft.DescribeTarget(p))
+}
+
+// traceIfCovered will log the message if it touches any of the registered trace
+// points. Additionally it returns any saved contexts by index for sending to
+// remote nodes. This typically applies to messages that the leader sends to the
+// followers.
+func (r *RaftTracer) traceIfCovered(m raftpb.Message) []kvserverpb.TracedEntry {
+	if len(m.Entries) == 0 {
+		return nil
+	}
+	minEntryIndex := kvpb.RaftIndex(m.Entries[0].Index)
+	maxEntryIndex := kvpb.RaftIndex(m.Entries[len(m.Entries)-1].Index)
+	var tracedEntries []kvserverpb.TracedEntry
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		// If the traced index is not in the range of the entries, we can skip
+		// it. We don't need to check each individual entry since they are
+		// contiguous.
+		if t.traced.Index < minEntryIndex || t.traced.Index > maxEntryIndex {
+			return true
+		}
+		tracedEntries = append(tracedEntries, t.traced)
+		// TODO(baptist): Not all the fields are relevant to log for all
+		// message types. Consider cleaning up what is logged.
+		t.logf(4,
+			"%s->%s %v Term:%d Log:%d/%d Range:%d-%d",
+			peer(m.From),
+			peer(m.To),
+			m.Type,
+			m.Term,
+			m.LogTerm,
+			m.Index,
+			minEntryIndex,
+			maxEntryIndex,
+		)
+		return true
+	})
+	return tracedEntries
+}
+
+// traceIfPast will log the message the message is past any registered tracing
+// points. It will additionally unregister traces that are no longer useful.
+// This call is for events that move the needle/watermark forward (e.g. the log
+// storage syncs), but don't have an exact range of entries affected. So, being
+// unable to match these events to entries exactly once, we instead check that
+// the watermark passed the entry. To protect against overly verbose logging, we
+// only allow MsgAppResp and MsgStorageAppendResp to be logged once per trace.
+func (r *RaftTracer) traceIfPast(m raftpb.Message) {
+	if m.Reject {
+		return
+	}
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		t.mu.Lock()
+		defer t.mu.Unlock()
+		switch m.Type {
+		case raftpb.MsgAppResp:
+			if kvpb.RaftIndex(m.Index) >= index && !t.mu.appRespSent[m.From] {
+				t.mu.appRespSent[m.From] = true
+				t.logf(4,
```

Comment:
Do we want to hold the `t.mu` mutex while logging? Could do the data structure updates with `t.mu` locked, remember whether we want to log the message, and log it at the end after having released the mutex.

Reply:
Updated. I'm not sure if it would have been a problem in practice, but I like the idea of minimizing the mutexes under lock.